src/nbl/asset/utils/IShaderCompiler.cpp

```diff
@@ -289,3 +219,116 @@ auto IShaderCompiler::CIncludeFinder::tryIncludeGenerators(const std::string& in
+    retVal.insert(retVal.end(), std::make_move_iterator(shadersBuffer.begin()), std::make_move_iterator(shadersBuffer.end()));
+    return core::make_smart_refctd_ptr<CVectorCPUBuffer<uint8_t, nbl::core::aligned_allocator<uint8_t>>>(std::move(retVal));
+}
+
+core::smart_refctd_ptr<IShaderCompiler::CCache> IShaderCompiler::CCache::deserialize(const std::span<const uint8_t> serializedCache)
+{
+    auto retVal = core::make_smart_refctd_ptr<CCache>();
+
+    // First get the size of the json in the buffer, stored in the first 8 bytes
+    const uint64_t* cacheStart = reinterpret_cast<const uint64_t*>(serializedCache.data());
+    uint64_t containerJsonSize = cacheStart[0];
+    // Next up get the json that stores the container data
+    std::span<const char> cacheAsChar = { reinterpret_cast<const char*>(serializedCache.data()), serializedCache.size() };
+    std::string_view containerJsonString(cacheAsChar.begin() + CONTAINER_JSON_SIZE_BYTES, cacheAsChar.begin() + CONTAINER_JSON_SIZE_BYTES + containerJsonSize);
+    json containerJson = json::parse(containerJsonString);
+    
+    // Now retrieve two vectors, one with the entries and one with the extra data to recreate the CPUShaders
+    std::vector<SEntry> entries;
+    std::vector<CPUShaderCreationParams> shaderCreationParams;
+    containerJson.at("entries").get_to(entries);
+    containerJson.at("shaderCreationParams").get_to(shaderCreationParams);
+
+    // We must now recreate the shaders, add them to each entry, then move the entry into the multiset
+    for (auto i = 0u; i < entries.size(); i++) {
+        // Create buffer to hold the code
+        auto code = core::make_smart_refctd_ptr<ICPUBuffer>(shaderCreationParams[i].codeByteSize);
+        // Copy the shader bytecode into the buffer
+        memcpy(code->getPointer(), serializedCache.data() + CONTAINER_JSON_SIZE_BYTES + containerJsonSize + shaderCreationParams[i].offset, shaderCreationParams[i].codeByteSize);
+        // Create the ICPUShader
+        auto value = core::make_smart_refctd_ptr<ICPUShader>(std::move(code), shaderCreationParams[i].stage, shaderCreationParams[i].contentType, std::move(shaderCreationParams[i].filepathHint));
+       
+        entries[i].value = std::move(value);
+
+        retVal->insert(std::move(entries[i]));
+    }
+
+    return retVal;
+}
+
+core::smart_refctd_ptr<IShaderCompiler::CCache> IShaderCompiler::CCache::deserialize(core::smart_refctd_ptr<const system::IFile> serializedCache) {
+    const void* fileBufferPointer = serializedCache->getMappedPointer();
+    assert(fileBufferPointer); // Check the file is readable
+    
+    size_t fileSize = serializedCache->getSize();
+    std::span cacheSpan{ reinterpret_cast<const uint8_t*>(fileBufferPointer), fileSize };
+    return IShaderCompiler::CCache::deserialize(std::span{ reinterpret_cast<const uint8_t*>(fileBufferPointer), fileSize });
```

Comment:
now you want a comment (if you want to leave as TODO) that this deserialization should work differently, basically that the `ICPUBuffer` of the spir-v code should be created differently

Reply:
Oh no this one does the same as the other one - it directly loads the whole cache at once from a File. The Cache that's streamed I'm planning to make it a different class, because some methods (such as the merging or the finding) need to be different. I might make those virtual, have the `CStreamedCache` extend `CCache` and override those methods

Reply:
ok, `CStreamedCache` would be a whole new task, which we don't have time for. So open a GH issue and we'll do it at some point later.

Thinking about it, we don't need a `deserialize` method from an `IFile` here.

To do such a thing "properly" we'd need an LZMA (7z) CArchive where:
- each shader is a virtual IFile thats LZMA compressed
- there's a JSON header IFile thats also LZMA compressed
- you decompress from LZMA on the fly when requesting a file

Reply:
write this up in a GH issue you'll open about a streaming cache